src/Symfony/Component/Finder/Tests/GitignoreTest.php

```diff
@@ -13,136 +13,313 @@
+            $this->assertMatchesRegularExpression(
+                $regex,
+                $matchingCase,
+                sprintf(
+                    "Failed asserting path:\n%s\nmatches gitignore patterns:\n%s",
+                    preg_replace('~^~m', '    ', $matchingCase),
+                    preg_replace('~^~m', '    ', $patterns)
+                )
+            );
         }
 
         foreach ($nonMatchingCases as $nonMatchingCase) {
-            $this->assertDoesNotMatchRegularExpression($regex, $nonMatchingCase, sprintf('Failed asserting path [%s] not matching gitignore patterns [%s] using regex [%s]', $nonMatchingCase, $patterns, $regex));
+            $this->assertDoesNotMatchRegularExpression(
+                $regex,
+                $nonMatchingCase,
+                sprintf("Failed asserting path:\n%s\nNOT matching gitignore patterns:\n%s",
+                    preg_replace('~^~m', '    ', $nonMatchingCase),
+                    preg_replace('~^~m', '    ', $patterns)
+                )
+            );
         }
     }
 
-    /**
-     * @return array return is array of
-     *               [
-     *               [
-     *               '', // Git-ignore Pattern
-     *               [], // array of file paths matching
-     *               [], // array of file paths not matching
-     *               ],
-     *               ]
-     */
     public function provider(): array
     {
-        return [
-            [
-                '
-                    *
-                    !/bin
-                    !/bin/bash
-                ',
-                ['bin/cat', 'abc/bin/cat'],
-                ['bin/bash'],
+        $cases = [
+            [
+                [''],
+                [],
+                ['a', 'a/b', 'a/b/c', 'aa', 'm.txt', '.txt'],
+            ],
+            [
+                ['a', 'X'],
+                ['a', 'a/b', 'a/b/c', 'X', 'b/a', 'b/c/a', 'a/X', 'a/X/y', 'b/a/X/y'],
+                ['A', 'x', 'aa', 'm.txt', '.txt', 'aa/b', 'b/aa'],
+            ],
+            [
+                ['/a', 'x', 'd/'],
+                ['a', 'a/b', 'a/b/c', 'x', 'a/x', 'a/x/y', 'b/a/x/y', 'd/', 'd/u', 'e/d/', 'e/d/u'],
+                ['b/a', 'b/c/a', 'aa', 'm.txt', '.txt', 'aa/b', 'b/aa', 'e/d'],
+            ],
+            [
+                ['a/', 'x'],
+                ['a/b', 'a/b/c', 'x', 'a/x', 'a/x/y', 'b/a/x/y'],
+                ['a', 'b/a', 'b/c/a', 'aa', 'm.txt', '.txt', 'aa/b', 'b/aa'],
+            ],
+            [
+                ['*'],
+                ['a', 'a/b', 'a/b/c', 'aa', 'm.txt', '.txt'],
+                [],
+            ],
+            [
+                ['/*'],
+                ['a', 'a/b', 'a/b/c', 'aa', 'm.txt', '.txt'],
+                [],
+            ],
+            [
+                ['/a', 'm/*'],
+                ['a', 'a/b', 'a/b/c', 'm/'],
+                ['aa', 'm', 'b/m', 'b/m/'],
```

Comment:
why `m/` and not `b/m/`?

Reply:
by spec, citing:

> If there is a separator at the beginning or middle (or both) of the pattern, then the pattern is relative to the directory level of the particular .gitignore file itself. Otherwise the pattern may also match at any level below the .gitignore level.